### ISSUE TITLE
Error executing lua /usr/share/nvim/runtime/lua/vim/lsp.lua:1301: table index is nil

### ISSUE THREAD

**Issue by jtw023** [ChristianChiarulli/LunarVim#293]
Whenever I try to write a file and stay in there with `:w` I've noticed that error popping up. From what I can tell, it has something to do with the buffer state since this also comes up:
`Error detected while processing BufferWritePre Autocommands for "*.py":`

I tried doing some research on both of those errors and the best source to come up seemed to be [here](https://github.com/neovim/nvim-lspconfig). Though unfortunately I couldn't figure it out.

That being said, this error hasn't affected anything in the way of functionality. Everything works exactly as I want it to so for me it's not a big deal if no one knows the solution off the top of their heads. Please don't spend too much time on this.

The output of `:LspInfo` is here:
![2021-04-22_19-54](https://user-images.githubusercontent.com/59935328/115811622-e3fe7400-a3a4-11eb-860b-58186886b712.png)

**Comment by rebuilt** [comment]
Initialize the working directory as a git repository so that efm will also attach to your buffer.

**Comment by rebuilt** [comment]
A more permanent solution would be to identify the current directory as root but the command I expect to do that doesn't work.  

```bash
root_dir = require'lspconfig'.util.root_pattern(".git","."),
```

**Comment by jtw023** [comment]
@rebuilt  I can't seem to figure out which file that `root_dir` variable should go in. Do I put it in the file of the lsp that i'm trying to use, the main init.lua file, or somewhere else?

**Comment by rebuilt** [comment]
~/.config/nvim/lua/lsp/efm-general-ls.lua around line 97

But it the line didn't work when I tried it.  I also tried changing the rootMarkers variable to set the root path but I couldn't make it work.  I'm not sure why.  

The only way it seems to work is to initialize a git repo.

**Comment by jtw023** [comment]
Hmm. I've also tried moving my `test.py` into a new directory called test and then copying a .git folder into that directory and the errors are still there. This one says something different though:
`Error executing vim.schedule lua callback: /usr/share/nvim/runtime/lua/vim/lsp/diagnostic.lua:1010: attempt to` and then it cuts off because wrap is not turned on.

**Comment by rebuilt** [comment]
Type :messages to see the whole error.

**Comment by jtw023** [comment]
Error executing vim.schedule lua callback: /usr/share/nvim/runtime/lua/vim/lsp/diagnostic.lua:1010: attempt to
 index local 'config' (a nil value)

**Comment by rebuilt** [comment]
It's surprising that running 'git init' at the command line didn't help efm identify the root directory.  What does 'git status' say?

**Comment by rebuilt** [comment]
Also, I think the first error and this most recent error are part of the same cycle of problems.  Python is set up to run linting.  But the linting isn't done by the python language server.  It's done by efm.  And in this case efm isn't set up correctly.  So every time you save a file efm tries and fails to lint it.

**Comment by jtw023** [comment]
`git status` returns all the files that the original directory was tracking and thinks they've all been deleted since it can't find them. And of course it shows the untracked `test.py`. 
This is so strange because literally everything else works. Formatting of lua files, formatting of python files, any new extensions that i've added. Maybe this is a legitimate bug and not just an issue on my local machine?

Speaking of, i've been looking at this [r lsp](https://github.com/neoclide/coc-r-lsp) but all that stuff is way over my head in terms of how to get it wired up without breaking anything else. I've been playing around with R scripting lately and I thought that might be a nice addition but, unfortunately, it doesn't look like i'm going to be the one to add it.

**Comment by jtw023** [comment]
I also found [this page here](https://github.com/neovim/nvim-lspconfig/tree/master/lua/lspconfig) which seems to already have it wired up but, again, this is too far over my head.

**Comment by rebuilt** [comment]
What's the content of your efm configuration file?  Maybe you have something unique in there

**Comment by jtw023** [comment]
Not sure if this is the right one but here are the contents of my `~/.config/nvim/lua/lsp/efm-general-ls.lua`(also, i'm using the yapf formatter. I didn't want to comment out or get rid of any of the other stuff because I thought something might break in terms of functionality):
~~~
-- Example configuations here: https://github.com/mattn/efm-langserver
-- TODO this file needs to be refactored eache lang should be it's own file
-- python
local python_arguments = {}

-- TODO replace with path argument
local flake8 = {
    LintCommand = "flake8 --ignore=E501 --stdin-display-name ${INPUT} -",
    lintStdin = true,
    lintFormats = {"%f:%l:%c: %m"}
}

local isort = {formatCommand = "isort --quiet -", formatStdin = true}

local yapf = {formatCommand = "yapf --quiet", formatStdin = true}
local black = {formatCommand = "black --quiet -", formatStdin = true}

if O.python.linter == 'flake8' then table.insert(python_arguments, flake8) end

if O.python.isort then table.insert(python_arguments, isort) end

if O.python.formatter == 'yapf' then
    table.insert(python_arguments, yapf)
elseif O.python.formatter == 'black' then
    table.insert(python_arguments, black)
end

-- lua
local lua_arguments = {}

local luaFormat = {
    formatCommand = "lua-format -i --no-keep-simple-function-one-line --column-limit=120",
    formatStdin = true
}

local lua_fmt = {
    formatCommand = "luafmt --indent-count 2 --line-width 120 --stdin",
    formatStdin = true
}

if O.lua.formatter == 'lua-format' then
  table.insert(lua_arguments, luaFormat)
elseif O.lua.formatter == 'lua-fmt' then
  table.insert(lua_arguments, lua_fmt)
end

-- sh
local sh_arguments = {}

local shfmt = {formatCommand = 'shfmt -ci -s -bn', formatStdin = true}

local shellcheck = {
    LintCommand = 'shellcheck -f gcc -x',
    lintFormats = {'%f:%l:%c: %trror: %m', '%f:%l:%c: %tarning: %m', '%f:%l:%c: %tote: %m'}
}

if O.sh.formatter == 'shfmt' then table.insert(sh_arguments, shfmt) end

if O.sh.linter == 'shellcheck' then table.insert(sh_arguments, shellcheck) end

-- tsserver/web javascript react, vue, json, html, css, yaml
local prettier = {formatCommand = "prettier --stdin-filepath ${INPUT}", formatStdin = true}
-- You can look for project scope Prettier and Eslint with e.g. vim.fn.glob("node_modules/.bin/prettier") etc. If it is not found revert to global Prettier where needed.
-- local prettier = {formatCommand = "./node_modules/.bin/prettier --stdin-filepath ${INPUT}", formatStdin = true}

local eslint = {
    lintCommand = "./node_modules/.bin/eslint -f unix --stdin --stdin-filename ${INPUT}",
    lintIgnoreExitCode = true,
    lintStdin = true,
    lintFormats = {"%f:%l:%c: %m"},
    formatCommand = "./node_modules/.bin/eslint --fix-to-stdout --stdin --stdin-filename=${INPUT}",
    formatStdin = true
}

local tsserver_args = {}

if O.tsserver.formatter == 'prettier' then table.insert(tsserver_args, prettier) end

if O.tsserver.linter == 'eslint' then table.insert(tsserver_args, eslint) end

-- local markdownlint = {
--     -- TODO default to global lintrc
--     -- lintcommand = 'markdownlint -s -c ./markdownlintrc',
--     lintCommand = 'markdownlint -s',
--     lintStdin = true,
--     lintFormats = {'%f:%l %m', '%f:%l:%c %m', '%f: %l: %m'}
-- }

local markdownPandocFormat = {formatCommand = 'pandoc -f markdown -t gfm -sp --tab-stop=2', formatStdin = true}

require"lspconfig".efm.setup {
    -- init_options = {initializationOptions},
    cmd = {"/home/jordan/.local/share/nvim/lspinstall/efm/efm-langserver"},
    init_options = {documentFormatting = true, codeAction = false},
    filetypes = {"lua", "python", "javascriptreact", "javascript", "sh", "html", "css", "json", "yaml", "markdown"},
    settings = {
        rootMarkers = {".git/"},
        languages = {
            python = python_arguments,
            lua = lua_arguments,
            sh = sh_arguments,
            javascript = tsserver_args,
            javascriptreact = tsserver_args,
            html = {prettier},
            css = {prettier},
            json = {prettier},
            yaml = {prettier},
            markdown = {markdownPandocFormat}
            -- javascriptreact = {prettier, eslint},
            -- javascript = {prettier, eslint},
            -- markdown = {markdownPandocFormat, markdownlint},
        }
    }
}

-- Also find way to toggle format on save
-- maybe this will help: https://superuser.com/questions/439078/how-to-disable-autocmd-or-augroup-in-vim
~~~

**Comment by jtw023** [comment]
I think I found the problem. `rootMarkers` is set to `.git/` and i'm assuming the forward slash indicates that it should be inside the actually git directory and not just a directory containing a git repo. 
I found that if I cd into a .git/ directory and then open a file inside there the warnings are gone. I can go as deep as I want to as long as i'm inside of the git directory.

**Comment by rebuilt** [comment]
Try creating a file called ROOT and adding "ROOT" to the list of rootMarkers.  I realize this is not an ideal solution but I haven't been able to make it identify the current working directory.  The trick of adding "." To the list doesn't work for this language server.

**Comment by jtw023** [comment]
I tried doing that in my efm-general-ls.lua file:
![2021-04-24_23-03](https://user-images.githubusercontent.com/59935328/115982604-61ff7e00-a551-11eb-8603-afb4e9762c71.png)
But now it led to this weird error here:
![2021-04-24_23-12](https://user-images.githubusercontent.com/59935328/115982806-a3dcf400-a552-11eb-867e-fde2491da853.png)

This is the current directory setup that test.py is in:
![2021-04-24_23-02](https://user-images.githubusercontent.com/59935328/115982654-a854dd00-a551-11eb-9669-75d115525379.png)
Honestly, I can live with the error. It's not affecting any functionality that i've noticed. As long as it stays that way and doesn't start affecting my productivity i'm fine with it.

**Comment by rebuilt** [comment]
Efm is responsible for linting your file.

**Comment by rebuilt** [comment]
Looks like the root_dir detection was fixed 14 days ago in this patch https://github.com/neovim/nvim-lspconfig/commit/9ce8ad469e7d9f9204a0315c260f918963d0df83 

If your version of lspconfig is older than 14 days try,

```bash
:PackerSync
```

**Comment by jtw023** [comment]
When I got on my computer and tried it out today it just worked for some reason. No errors whatsoever. Even without the root file fix.

**Comment by rebuilt** [comment]
Were you restarting neovim after making changes to your configs?   It has to restart to load new configurations.

**Comment by jtw023** [comment]
@rebuilt Absolutely! Today i've even turned my whole system off for a few hours and then came back to this question. Everything is working perfectly. I wonder if there was an issue in the neovim-git aur package that was causing the problem. All I know is that I updated my system this morning and suddenly everything was working perfectly as long as it's in a git directory.

**Comment by jtw023** [comment]
@rebuilt The error is back and I have since figured out what's up. In my ~$USER/.config/nvim/lv-settings.lua I had a line `O.python.autoformat = true` which was causing the program to try to autoformat my code everytime I saved a python file. I've since turned that off and all errors are completely gone. So essentially, lunarvim is trying to autoformat my files without a formatter installed apparently. 

I have created another [issue](https://github.com/ChristianChiarulli/LunarVim/issues/341) for this now that i've diagnosed the exact problem.
